src/wallet/wallet.h

```diff
@@ -1544,5 +1542,162 @@ class AddSpendingKeyToWallet : public boost::static_visitor<KeyAddResult>
     KeyAddResult operator()(const libzcash::InvalidEncoding& no) const;
 };
 
+struct LogTx
+{
+    std::map<std::string, CAmount> balances;
+    enum TransactionStateType { TX_GENERATED, TX_MINED };
+    std::map<uint256, TransactionStateType> txstate;
+
+    std::string ValueFromAmount(const CAmount& amount)
+    {
+        bool sign = amount < 0;
+        int64_t n_abs = (sign ? -amount : amount);
+        int64_t quotient = n_abs / COIN;
+        int64_t remainder = n_abs % COIN;
+        return strprintf("%s%d.%08d", sign ? "-" : "", quotient, remainder);
+    }
+
+    void LogBalance(const std::string& address, const CAmount& amount, const bool sum = false)
+    {
+        if (balances.count(address) != 0) {
+            if (balances.at(address) != amount && amount != 0) {
+                const CAmount total_amount_prev = balances.at(address);;
+                CAmount total_amount_now;
+                if (!sum)
+                    total_amount_now = amount;
+                else
+                    total_amount_now = balances.at(address)+amount;
+
+                LogPrint("balanceunsafe", "Balance changed in address %s from %s to %s\n",
+                    address, ValueFromAmount(total_amount_prev), ValueFromAmount(total_amount_now));
+                balances[address] = total_amount_now;
+            }
+        }
+        else {
+            LogPrint("balanceunsafe", "Balance changed in address %s from %s to %s\n",
+                address, ValueFromAmount(0), ValueFromAmount(amount));
+            balances[address] = amount;
+        }
+    }
+
+    TransactionStateType LogState(const uint256 &hashTx)
+    {
+        TransactionStateType state;
+        if (txstate.count(hashTx) == 0) {
```

Comment:
This is the only place that I can find `txstate` being read from, and it's equivalent to `pwalletMain->mapWallet.count(hashTx)`. Were you planning to use `txstate` for anything else? It feels like we should be deriving this state from the wallet instead of maintaining a parallel memory-only structure.

Reply:
https://github.com/zcash/zcash/pull/4423/commits/fd6371aafea69cdb19c4fa770cd2a7d2ee8cea5b